# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/ipc/jsactor/JSProcessActorChild.cpp
# Commit: 7229926a45c2
# Full Hash: 7229926a45c2e43136c7f8f597a92776658edbd5
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-10-14 09:52:12
# Description:
#   Bug 1660539 - Make StructuredCloneData args optional in JSActor RawMessage, r=kmag
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D93122
# ==============================================================================

diff -r 1aa2dc4b280e -r 7229926a45c2 dom/ipc/jsactor/JSProcessActorChild.cpp
--- a/dom/ipc/jsactor/JSProcessActorChild.cpp	Tue Oct 13 16:57:32 2020 +0000
+++ b/dom/ipc/jsactor/JSProcessActorChild.cpp	Tue Oct 13 17:43:56 2020 +0000
@@ -28,17 +28,22 @@
   return JSProcessActorChild_Binding::Wrap(aCx, this, aGivenProto);
 }
 
-void JSProcessActorChild::SendRawMessage(const JSActorMessageMeta& aMeta,
-                                         ipc::StructuredCloneData&& aData,
-                                         ipc::StructuredCloneData&& aStack,
-                                         ErrorResult& aRv) {
+void JSProcessActorChild::SendRawMessage(
+    const JSActorMessageMeta& aMeta, Maybe<ipc::StructuredCloneData>&& aData,
+    Maybe<ipc::StructuredCloneData>&& aStack, ErrorResult& aRv) {
   if (NS_WARN_IF(!CanSend() || !mManager || !mManager->GetCanSend())) {
     aRv.ThrowInvalidStateError("JSProcessActorChild cannot send at the moment");
     return;
   }
 
-  if (NS_WARN_IF(
-          !AllowMessage(aMeta, aData.DataLength() + aStack.DataLength()))) {
+  size_t length = 0;
+  if (aData) {
+    length += aData->DataLength();
+  }
+  if (aStack) {
+    length += aStack->DataLength();
+  }
+  if (NS_WARN_IF(!AllowMessage(aMeta, length))) {
     aRv.ThrowDataCloneError(
         nsPrintfCString("JSProcessActorChild serialization error: data too "
                         "large, in actor '%s'",
@@ -56,16 +61,26 @@
     return;
   }
 
-  ClonedMessageData msgData;
-  ClonedMessageData stackData;
-  if (NS_WARN_IF(
-          !aData.BuildClonedMessageDataForChild(contentChild, msgData)) ||
-      NS_WARN_IF(
-          !aStack.BuildClonedMessageDataForChild(contentChild, stackData))) {
-    aRv.ThrowDataCloneError(nsPrintfCString(
-        "JSProcessActorChild serialization error: cannot clone, in actor '%s'",
-        PromiseFlatCString(aMeta.actorName()).get()));
-    return;
+  // Cross-process case - send data over ContentChild to other side.
+  Maybe<ClonedMessageData> msgData;
+  if (aData) {
+    msgData.emplace();
+    if (NS_WARN_IF(
+            !aData->BuildClonedMessageDataForChild(contentChild, *msgData))) {
+      aRv.ThrowDataCloneError(
+          nsPrintfCString("JSProcessActorChild serialization error: cannot "
+                          "clone, in actor '%s'",
+                          PromiseFlatCString(aMeta.actorName()).get()));
+      return;
+    }
+  }
+
+  Maybe<ClonedMessageData> stackData;
+  if (aStack) {
+    stackData.emplace();
+    if (!aStack->BuildClonedMessageDataForChild(contentChild, *stackData)) {
+      stackData.reset();
+    }
   }
 
   if (NS_WARN_IF(!contentChild->SendRawMessage(aMeta, msgData, stackData))) {